File: Examples/objective-c/Notification/Notification/Estimote/NotificationManager.m
//
// Please report any problems with this app template to contact@estimote.com
//

#import "NotificationManager.h"
#import <UIKit/UIKit.h>
#import <EstimoteSDK/EstimoteSDK.h>


@interface NotificationManager () <ESTMonitoringV2ManagerDelegate>

@property (nonatomic) ESTMonitoringV2Manager *monitoringManager;
@property (nonatomic) NSString *enterMessage;
@property (nonatomic) NSString *exitMessage;

@end

@implementation NotificationManager

- (instancetype)init {
    self = [super init];
    if (self) {
        _monitoringManager = [[ESTMonitoringV2Manager alloc] initWithDesiredMeanTriggerDistance:1.5
                                                                                       delegate:self];
    }
    
    return self;
}

- (void)enableNotificationsForDeviceIdentifier:(NSString *)beaconIdentifier
                                  enterMessage:(NSString *)enterMessage
                                   exitMessage:(NSString *)exitMessage {
    // Set up local notifications
    UIUserNotificationType notificationTypes = (UIUserNotificationTypeAlert | UIUserNotificationTypeSound);
    UIUserNotificationSettings *notificationSettings = [UIUserNotificationSettings settingsForTypes:notificationTypes
                                                                                         categories:nil];
    [[UIApplication sharedApplication] registerUserNotificationSettings:notificationSettings];
    
    // Remember messages to show
    self.enterMessage = enterMessage;
    self.exitMessage = exitMessage;
    
    // Start monitoring
    [self.monitoringManager startMonitoringForIdentifiers:@[beaconIdentifier]];
}

#pragma mark - Private

- (void)showNotificationWithMessage:(NSString *)message {
    UILocalNotification *notification = [UILocalNotification new];
    notification.alertBody = message;
    notification.soundName = UILocalNotificationDefaultSoundName;
    [[UIApplication sharedApplication] presentLocalNotificationNow:notification];
}

- (NSString *)nameForState:(ESTMonitoringState)state {
    switch (state) {
        case ESTMonitoringStateUnknown:
            return @"Unknown";
        case ESTMonitoringStateInsideZone:
            return @"Inside";
        case ESTMonitoringStateOutsideZone:
            return @"Outside";
    }
}

#pragma mark ESTMonitoringV2ManagerDelgate

- (void)monitoringManagerDidStart:(ESTMonitoringV2Manager *)manager {
    NSLog(@"Started monitoring successfully.");
}

- (void)monitoringManager:(ESTMonitoringV2Manager *)manager didDetermineInitialState:(ESTMonitoringState)state forBeaconWithIdentifier:(NSString *)identifier {
    NSLog(@"Determined initial state for %@ - %@.", identifier, [self nameForState:state]);
}

- (void)monitoringManager:(ESTMonitoringV2Manager *)manager didEnterDesiredRangeOfBeaconWithIdentifier:(NSString *)identifier {
    NSLog(@"Entered desired range of %@.", identifier);
    [self showNotificationWithMessage:self.enterMessage];
}

- (void)monitoringManager:(ESTMonitoringV2Manager *)manager didExitDesiredRangeOfBeaconWithIdentifier:(NSString *)identifier {
    NSLog(@"Exited desired range of %@.", identifier);
    [self showNotificationWithMessage:self.exitMessage];
}

- (void)monitoringManager:(ESTMonitoringV2Manager *)manager didFailWithError:(NSError *)error {
    NSLog(@"Monitoring failed. Error: %@.", error);
}

@end

File: Examples/swift/Notification/Notification/Estimote/NotificationManager.swift
//
// Please report any problems with this app template to contact@estimote.com
//

import Foundation
import UIKit

extension NotificationManager {
    fileprivate struct Parameters {
        
        /// Desired distance from a beacon from which enter/exit events should be triggered, in meters.
        static let desiredZoneRadius = 2.0
    }
}

class NotificationManager: NSObject {
    
    fileprivate lazy var monitoringManager: ESTMonitoringV2Manager = ESTMonitoringV2Manager(desiredMeanTriggerDistance: Parameters.desiredZoneRadius, delegate: self)
    fileprivate var enterMessage: String?
    fileprivate var exitMessage: String?
    
    internal func enableNotifications(deviceIdentifier: String, enterMessage: String, exitMessage: String) {
        
        // Set up local notifications
        let notificationSettings = UIUserNotificationSettings(types: [.alert, .sound], categories: nil)
        UIApplication.shared.registerUserNotificationSettings(notificationSettings)
        
        // Remember messages to show on enter/exit events
        self.enterMessage = enterMessage
        self.exitMessage = exitMessage
        
        self.monitoringManager.startMonitoring(forIdentifiers: [deviceIdentifier])
    }
    
    fileprivate func presentNotification(message: String) {
        let notification = UILocalNotification()
        notification.alertBody = message
        notification.soundName = UILocalNotificationDefaultSoundName
        UIApplication.shared.presentLocalNotificationNow(notification)
    }
    
    fileprivate func name(_ state: ESTMonitoringState) -> String {
        switch state {
        case .unknown: return "Unknown"
        case .insideZone: return "Inside"
        case .outsideZone: return "Outside"
        }
    }
}

extension NotificationManager: ESTMonitoringV2ManagerDelegate {
    
    func monitoringManagerDidStart(_ manager: ESTMonitoringV2Manager) {
        print("Monitoring started successfully.")
    }
    
    func monitoringManager(_ manager: ESTMonitoringV2Manager, didFailWithError error: Error) {
        print("Monitoring failed: \(error.localizedDescription)")
    }
    
    func monitoringManager(_ manager: ESTMonitoringV2Manager, didDetermineInitialState state: ESTMonitoringState, forBeaconWithIdentifier identifier: String) {
        print("Determined initial state: for \(identifier) - \(self.name(state)).")
    }
    
    func monitoringManager(_ manager: ESTMonitoringV2Manager, didEnterDesiredRangeOfBeaconWithIdentifier identifier: String) {
        print("Entered range of \(identifier).")
        self.presentNotification(message: self.enterMessage!)
    }
    
    func monitoringManager(_ manager: ESTMonitoringV2Manager, didExitDesiredRangeOfBeaconWithIdentifier identifier: String) {
        print("Exited range of \(identifier).")
        self.presentNotification(message: self.exitMessage!)
    }
}
